### PR TITLE
Use bucket region cache to minimize getLocation requests

### PR DESCRIPTION
**Pull request by krisis**

### CODE REVIEW

**Comment by donatello** [comment]
Region lookup is better done in `buildRequest`, so it applies to all APIs rather than just PutObject.

**Comment by krisis** [comment]
A local functional test on aws s3 with us-west-1 set as region in ConnectInfo worked. Suggestions on unit/functional tests welcome.

File: src/Network/Minio/Data.hs [hunk 1]
```diff
@@ -5,19 +5,28 @@ import           Control.Monad.Base
...
 import           Lib.Prelude
 
-import           Network.Minio.Errors
-import           Network.Minio.Utils
+
+-- | awsRegionMap - library constant
+awsRegionMap :: Map.Map Text Text
+awsRegionMap = Map.fromList [
+    ("us-east-1", "s3.amazonaws.com")
+  , ("us-west-1", "s3-us-west-1.amazonaws.com")
```
Comment (donatello) [inline src/Network/Minio/Data.hs]: Now that it works, all the regions can be added.

File: src/Network/Minio/Data.hs [hunk 2]
```diff
@@ -45,6 +55,21 @@ awsCI = def {
...
+-- This is for users who don't want minio-hs discovering region of a
+-- bucket if not known.
+awsWithRegion :: Region -> Bool -> ConnectInfo
+awsWithRegion region autoDiscoverRegion =
+  let host = maybe "s3.amazonaws.com" identity $
+             Map.lookup region awsRegionMap
+  in awsCI {
+      connectHost = host
+    , connectRegion = region
+    , connectAutoDiscoverRegion = autoDiscoverRegion
```
Comment (donatello) [inline src/Network/Minio/Data.hs]: No need to set secure=On?

Comment (krisis) [inline src/Network/Minio/Data.hs]: awsCI has connectIsSecure set to True

Comment (donatello) [inline src/Network/Minio/Data.hs]: :+1:

**Pull request merged**